account_payment_line_cancel/models/account_payment_line.py

```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Compassion CH (http://www.compassion.ch)
+# @author: Marco Monzione <marco.mon@windowslive.com>, Emanuel Cino
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, _
+
+
+class AccountCancelPayment(models.Model):
+    _inherit = 'account.payment.line'
+
+    cancel_reason = fields.Char()
+
+    def cancel_line(self):
+        """
+         This method proceed to call private methods to remove payment lines
+         from payment order
+
+        A cancel reason can be put in the cancel_reason field.
+
+        :return: True
+        """
+        # Retrieve all account_move_line(s) associated with the previous
+        # account_payment_line(s).
+        all_account_move_lines = self.mapped('move_line_id')
+
+        # Add the message to the invoice and to the payment order
+        self._post_cancel_message()
+
+        # Unreconcile transfer journal entries
+        full_reconcile_ids = all_account_move_lines.mapped(
+            'full_reconcile_id.id')
+        payment_orders = self.mapped('order_id')
+        if full_reconcile_ids:
+            # Retrieve the counterparts of the previous move_lines.
+            # The counter part should always be unique.
+            account_move_line_counterpart = self.env[
+                'account.move.line'].search([
+                    ('full_reconcile_id', 'in', full_reconcile_ids),
+                    ('id', 'not in', all_account_move_lines.ids),
+                    ('move_id.payment_order_id', 'in', payment_orders.ids)])
+
+            # All the move lines with the same reconcile id are unreconciled.
+            account_move_line_counterpart.remove_move_reconcile()
```

Comment:
Are you sure this should be done automatically? It's very risky IMO

Reply:
If we don't do it, the cancel method leaves move lines that are no more relevant, and the accounting people need to be aware of that. I don't think it's risky, as we unreconcile and remove the payment lines that were created especially for the payment line.